pkg/containers/containers.go

```diff
@@ -182,62 +202,98 @@ func (c *Containers) populate() error {
 // saving container information in Containers CgroupInfo map.
 // NOTE: ALL given cgroup dir paths are stored in CgroupInfo map.
 func (c *Containers) CgroupUpdate(cgroupId uint64, path string, ctime time.Time) (CgroupInfo, error) {
-	containerId, runtime := getContainerIdFromCgroup(path)
+	containerId, containerRuntime := getContainerIdFromCgroup(path)
+	container := cruntime.ContainerMetadata{
+		ContainerId: containerId,
+	}
 
 	info := CgroupInfo{
-		Path:        path,
-		ContainerId: containerId,
-		Runtime:     runtime,
-		Ctime:       ctime,
+		Path:      path,
+		Container: container,
+		Runtime:   containerRuntime,
+		Ctime:     ctime,
 	}
 
 	c.mtx.Lock()
 	c.cgroups[uint32(cgroupId)] = info
 	c.mtx.Unlock()
 
+	err := c.enrichCgroupInfo(cgroupId)
+
+	//don't fail on enrichment but log to stderr
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "failed to enrich container data: %v", err)
+	}
+
 	return info, nil
 }
 
+// enrichCgroupInfo checks for a given cgroupId if it is relevant to some running container
+// it then calls the runtime info service to gather additional data from the container's runtime
+func (c *Containers) enrichCgroupInfo(cgroupId uint64) error {
+	c.mtx.RLock()
+	info := c.cgroups[uint32(cgroupId)]
+	c.mtx.RUnlock()
+	containerId := info.Container.ContainerId
+	runtime := info.Runtime
+
+	if containerId != "" && runtime != cruntime.Unknown {
+		metadata, err := c.enricher.Get(containerId, runtime, context.Background())
+
+		//if enrichment fails, just return early
+		if err != nil {
+			return err
+		}
+
+		info.Container = metadata
+		c.mtx.Lock()
+		c.cgroups[uint32(cgroupId)] = info
```

Comment:
Shouldn't you check if `c.cgroups[uint32(cgroupdId)]` still exists when holding the mutex again ? It could have been removed after you read it for the first time AND you would be reintroducing it here (unsure if it could happen in practice, but there is a possibility in theory).

Reply:
Fair point, I can add that check. Should we return an error in such a case?

Reply:
Perhaps just ignore ? Its like "we don't need enrichment any longer".. unless you can think of something else.

Reply:
Reasonable to me, ignoring was also what I had in mind.

Reply:
Done.